proof/refine/ARM/Reply_R.thy

```diff
@@ -33,6 +54,17 @@ lemma replyTCB_update_reply_projs[wp]:
      apply (clarsimp simp: ext opt_map_def obj_at'_def projectKO_eq)+
   done
 
+lemma updateReply_reply_projs:
+  "\<lbrace>\<lambda>s. \<forall>ko. ko_at' ko rptr s \<longrightarrow>
+       P (\<lambda>a. if a = rptr then replyNext_of (f ko) else replyNexts_of s a)
+         (\<lambda>a. if a = rptr then replyPrev (f ko) else replyPrevs_of s a)
+         (\<lambda>a. if a = rptr then replyTCB (f ko) else replyTCBs_of s a)
+         (\<lambda>a. if a = rptr then replySc (f ko) else replySCs_of s a)\<rbrace>
```

Comment:
I am not entirely sure, but do we want to write this with implications or with `if _ then _ else _`?

Reply:
I'm not sure either, but I just moved this lemma and extended it to all of the projections.

Reply:
I think it's good this way: it's very easy to read (easier than implications and conjunctions) and I don't think it's any harder to manage otherwise.

Reply:
I remember some discussions that went like wpsimp being too eager to split if-statements, leading to preferring implication over if-then-else when expressions are small (I might be misremembering the details).
Having said that, this is almost similar to `fun_upd`, so probably we don’t need to worry too much anyway.

Reply:
@lsf37  might remember more, maybe?